### PR TITLE
prov/efa:  Fix fi_sendmsg/fi_writemsg with FI_INJECT

### PR DESCRIPTION
**Pull request by jiaxiyan**
Currently EFA provider doesn't handle the *msg call with FI_INJECT flags correctly. It allows user to pass in a buffer with valid MR while doing inject - outbound buffer should be reusable immediately after function returns.
That being said, without the inline HW support, EFA provider should do a copy from the user buffer to the internal bounce buffer even if application provides a valid MR for the buffer.
The inject rdma write path assumes the outbound buffer is always FI_HMEM_SYSTEM, but it doesn't have to.

### CODE REVIEW

File: prov/efa/src/rdm/efa_rdm_pke_utils.c [hunk 1]
```diff
@@ -82,13 +82,17 @@ ssize_t efa_rdm_pke_init_payload_from_ope(struct efa_rdm_pke *pke,
...
+	 * 2. data to be send is in 1 iov, because device only support 2 iov,
+	 * and we use 1st iov for header.
+	 * 3. FI_INJECT transfer flag is not set. FI_INJECT requires outbound
+	 * buffer to be reusable immediately after function returns, so without
+	 * the inline HW support, a copy is needed.
 	 */
 	if (efa_can_access_memory &&
-	    (tx_iov_offset + data_size <= ope->iov[tx_iov_index].iov_len)) {
+	    (tx_iov_offset + data_size <= ope->iov[tx_iov_index].iov_len) &&
+	    !(ope->fi_flags & FI_INJECT)) {
```
Comment (shijin-aws) [inline prov/efa/src/rdm/efa_rdm_pke_utils.c:95]: I don't think this is correct.

FI_INJECT can still avoid copy when the buffer is on host and size is within inline. It seems we need some more refactor here.

File: prov/efa/src/rdm/efa_rdm_pke_utils.c [hunk 2]
```diff
@@ -100,6 +100,37 @@ ssize_t efa_rdm_pke_init_payload_from_ope(struct efa_rdm_pke *pke,
...
+ *
+ * @param[in]		iov_mr  memory descriptors of the user's data buffer.
+ * @param[in,out]	pke	    packet entry. Header must have been set when the function is called.
+ * @param[in]		ope		operation entry that has user buffer information.
+ * @param[in]		payload_offset	the data offset in reference to pkt_entry->wiredata.
+ * @param[in]		segment_offset	the data offset in reference to user's buffer
+ * @param[in]		data_size	length of the data to be set up.
+ * @return   		length of data copied.
+ */
+size_t efa_rdm_copy_from_hmem_iov(struct efa_mr *iov_mr,
```
Comment (shijin-aws) [inline prov/efa/src/rdm/efa_rdm_pke_utils.c:125]: I prefer make it as a inline function in `efa_rdm_pke_utils.h`

Comment (shijin-aws) [reply prov/efa/src/rdm/efa_rdm_pke_utils.c:125]: also name it as `efa_rdm_pke_copy_from_hmem_iov`

File: prov/efa/src/rdm/efa_rdm_pke_utils.c [hunk 3]
```diff
@@ -61,34 +62,36 @@ ssize_t efa_rdm_pke_init_payload_from_ope(struct efa_rdm_pke *pke,
...
 	/*
 	 * Copy can be avoid if the following 2 conditions are true:
-	 * 1. EFA can directly access the memory
-	 * 2. data to be send is in 1 iov, because device only support 2 iov, and we use
-	 *    1st iov for header.
+	 * 1. data to be send is in 1 iov, because device only support 2 iov,
+	 * and we use 1st iov for header.
+	 * 2. EFA can directly access the memory
+	 * Note that FI_INJECT requires outbound buffer to be reusable
+	 * immediately after function returns, so without the inline HW support,
```
Comment (shijin-aws) [inline prov/efa/src/rdm/efa_rdm_pke_utils.c:90]: Can you rephrase the `without the inline HW support,` to `unless using inline`

File: prov/efa/src/rdm/efa_rdm_pke_utils.c [hunk 4]
```diff
@@ -61,34 +62,36 @@ ssize_t efa_rdm_pke_init_payload_from_ope(struct efa_rdm_pke *pke,
 	assert(tx_iov_index < ope->iov_count);
 	assert(tx_iov_offset < ope->iov[tx_iov_index].iov_len);
 	iov_mr = ope->desc[tx_iov_index];
-
+	p2p_available = false;
+	inline_send = false;
 	/* When using EFA device, EFA device can access memory that
```
Comment (shijin-aws) [inline prov/efa/src/rdm/efa_rdm_pke_utils.c:67]: Since we are removing `efa_can_access_memory`, I think we can remove these comments, just explain how copy can be removed based on the p2p_available and inline_send status later.

**Review by shijin-aws — APPROVED** [review]
LGTM, some nits on the comments

File: prov/efa/src/rdm/efa_rdm_pke_utils.c (hunk 3 above)
Comment (shijin-aws) [reply prov/efa/src/rdm/efa_rdm_pke_utils.c:90]: Also can we change `EFA can directly access the memory` to `p2p is available or using inline` here

File: prov/efa/src/rdm/efa_rdm_pke_utils.h [hunk 5]
```diff
@@ -58,6 +58,42 @@ size_t efa_rdm_pke_get_segment_offset(struct efa_rdm_pke *pke)
...
+ *
+ * @param[in]		iov_mr  memory descriptors of the user's data buffer.
+ * @param[in,out]	pke	    packet entry. Header must have been set when the function is called.
+ * @param[in]		ope		operation entry that has user buffer information.
+ * @param[in]		payload_offset	the data offset in reference to pkt_entry->wiredata.
+ * @param[in]		segment_offset	the data offset in reference to user's buffer
+ * @param[in]		data_size	length of the data to be set up.
+ * @return   		length of data copied.
+ */
+inline size_t
```
Comment (shijin-aws) [inline prov/efa/src/rdm/efa_rdm_pke_utils.h:72]: I missed it earlier: it should be `static inline size_t`

Comment (jiaxiyan) [reply prov/efa/src/rdm/efa_rdm_pke_utils.h:72]: Can I call it from efa_rdm_ope.c if it is static?

**Comment by shijin-aws** [comment]
bot:aws:retest
